core/utils/historical-data/src/linear.rs

```diff
@@ -0,0 +1,447 @@
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Linear historical data.
+
+#[cfg(not(feature = "std"))]
+use rstd::{vec::Vec, vec};
+use rstd::marker::PhantomData;
+use rstd::borrow::Cow;
+use crate::HistoricalValue;
+
+
+/// Array like buffer for in memory storage.
+/// By in memory we expect that this will
+/// not required persistence and is not serialized.
+#[cfg(not(feature = "std"))]
+pub(crate) type MemoryOnly<V, I> = Vec<HistoricalValue<V, I>>;
+
+/// Array like buffer for in memory storage.
+/// By in memory we expect that this will
+/// not required persistence and is not serialized.
+#[cfg(feature = "std")]
+pub(crate) type MemoryOnly<V, I> = smallvec::SmallVec<[HistoricalValue<V, I>; ALLOCATED_HISTORY]>;
+
+/// Size of preallocated history per element.
+/// Currently at two for committed and prospective only.
+/// It means that using transaction in a module got a direct allocation cost.
+#[cfg(feature = "std")]
+const ALLOCATED_HISTORY: usize = 2;
+
+#[derive(Debug, Clone)]
+#[cfg_attr(any(test, feature = "test"), derive(PartialEq))]
+/// Arraylike buffer with in place byte data.
+/// Can be written as is in underlying
+/// storage.
+/// Could be extended to direct access memory too.
+pub struct Serialized<'a, F>(SerializedBuff<'a>, PhantomData<F>);
+
+#[derive(Debug)]
+#[cfg_attr(any(test, feature = "test"), derive(PartialEq))]
+enum SerializedBuff<'a> {
+	Cow(Cow<'a, [u8]>),
+	Mut(&'a mut Vec<u8>),
+}
+
+impl<'a> SerializedBuff<'a> {
+	pub fn to_mut(&mut self) -> &mut Vec<u8> {
+		match self {
+			SerializedBuff::Cow(c) => c.to_mut(),
+			SerializedBuff::Mut(m) => m,
+		}
+	}
+	pub fn into_owned(self) -> Vec<u8> {
+		match self {
+			SerializedBuff::Cow(c) => c.into_owned(),
+			SerializedBuff::Mut(m) => m.clone(),
+		}
+	}
+}
+
+impl<'a> rstd::ops::Deref for SerializedBuff<'a> {
+	type Target = [u8];
+	fn deref(&self) -> &Self::Target {
+		match self {
+			SerializedBuff::Cow(c) => c.deref(),
+			SerializedBuff::Mut(m) => m.deref(),
+		}
+	}
+}
+
+impl<'a> rstd::ops::DerefMut for SerializedBuff<'a> {
+	fn deref_mut(&mut self) -> &mut Self::Target {
+		&mut self.to_mut()[..]
+	}
+}
+
+impl<'a> Clone for SerializedBuff<'a> {
+	fn clone(&self) -> Self {
+		match self {
+			SerializedBuff::Cow(c) => SerializedBuff::Cow(c.clone()),
+			SerializedBuff::Mut(m) => {
+				let m: Vec<u8> = (*m).clone();
+				SerializedBuff::Cow(Cow::Owned(m))
+			}
+		}
+	}
+}
+
+/// Serialized specific behavior.
+pub trait SerializedConfig {
+	/// encoded empty slice
+	fn empty() -> &'static [u8];
+	/// size at start for encoding version.
```

Comment:
does that means length used at start of value to encode the version ?

at start of what ?